### PR TITLE
Type hinting

### PR DESCRIPTION
**Pull request by lk-geimfari**
What was done:

- [x] Corrected most annotations.
- [x] Removed solutions added to support Python 3.3.
- [x] Added more types to `typing.py`.
- [x] Removed Python 3.3/3.4 from classifiers in `setup.py`.
- [x] Updated Python version badge.
- [x] Other minor changes and improvements.

Summary:

At this moment i have confused by few messages of `mypy`. I have created file `MYPY_REPORT.txt` and i hope for help with this issues.

Issue:  #258.

### CODE REVIEW

**Comment by lk-geimfari** [comment]
@sobolevn I hope for your help.

File: MYPY_REPORT.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+mimesis/providers/units.py:299: error: Argument 1 to "choice" of "Random" has incompatible type "List[str]"; expected "Sequence[Dict[str, List[str]]]"
+mimesis/providers/generic.py:71: error: Item "function" of "Union[Callable[..., Any], Any]" has no attribute "Meta"
```
Comment (sobolevn) [inline MYPY_REPORT.txt]: You are using a class, not function.

Comment (sobolevn) [reply MYPY_REPORT.txt]: `Provider` should be a protocol: http://mypy.readthedocs.io/en/latest/class_basics.html#simple-user-defined-protocols

Comment (lk-geimfari) [reply MYPY_REPORT.txt]: @sobolevn Should i create `protoclos.py` or just create it in `typing.py`?

Comment (sobolevn) [reply MYPY_REPORT.txt]: IMHO stick to `typing.py`

Comment (lk-geimfari) [reply MYPY_REPORT.txt]: @sobolevn Seems like i have not `typing_extensions` module in my version of Python 3.5.2 on Ubuntu 16.04:

Comment (sobolevn) [reply MYPY_REPORT.txt]: https://github.com/python/typing/issues/467

Comment (lk-geimfari) [reply MYPY_REPORT.txt]: So, it's mean that we should add third-party dependencies and i'm not sure that it's worth it. I will just ignore places where we need Protocol.

File: docs/conf.py [hunk 2]
```diff
@@ -33,6 +33,7 @@
 # ones.
 extensions = [
     'sphinx.ext.autodoc',
+    'sphinx_autodoc_typehints'
```
Comment (lk-geimfari) [inline docs/conf.py]: Trailing comma.

**Comment by bjd2385** [comment]
I'm going to propose the following stylistic guidelines while contributing type annotations.

- Follow PEP 8.
- Use `mimesis.typing` for defining Mimesis-only type variables, such as `TimeStamp`, `JSON`, etc. It's actually rather annoying to have to keep typing `types.Any`, when we should be importing them directly from Python's `typing` module at the top of the file, i.e. `from typing import Any`.
- For type checking I'm using the following command (are there any additional restrictions you'd like to impose?): `mypy --strict --strict-optional --check-untyped-defs`

Another thing I'm seeing over and over again is the use of spaces before default arguments, e.g. the following, which can be found in `helpers.py`:

> def randints(self, amount: int = 3, a: int = 1, b: int = 100) -> list:

This should instead be, per PEP 8 guidelines,

> def randints(self, amount: int =3, a: int =1, b: int =100) -> list:

I know it's a small thing to pick lol, it's just one of my pet peaves :P

**Comment by lk-geimfari** [comment]
@bjd2385 I agree that spaces in arguments is not cool, but only when arguments have not annotations.  But in our case this is what PyCharm do automatically and i like it, because it's more readable.

```
def randints(self, amount: int = 3, a: int = 1, b: int = 100) -> list:
    ...
```

**Comment by lk-geimfari** [comment]
@bjd2385 As for the rest of your comments, I agree with you and i will fix everything.

**Comment by lk-geimfari** [comment]
@bjd2385 Also i recommend you look at this [PEP](https://www.python.org/dev/peps/pep-3107/#syntax).

**Pull request merged**